File: Engine/Android/core/src/test/kotlin/despia/engine/RefConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The `ref=` conformance runner - executes OpenSource/Conformance/input/ref.json through THIS
 * runtime's StackRef + RefRegistry. The TS twin (@despia-native/kernel ref.ts) and the Swift twin run
 * the SAME file, so a recycled list row cannot keep a ref alive on one renderer and drop it on
 * another.
 */
class RefConformanceTest {

    private fun corpusFile(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val c = File(dir, "OpenSource/Conformance/input/ref.json")
            if (c.isFile) return c
            dir = dir.parentFile ?: error("OpenSource/Conformance/input/ref.json not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun section(name: String): List<Map<String, Any?>> {
        val root = json(corpusFile().readText()).foundationValue as? Map<String, Any?>
            ?: error("ref.json: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "ref.json version")
        return (root[name] as? List<Map<String, Any?>>)?.also {
            assertTrue(it.isNotEmpty(), "$name must not be empty")
        } ?: error("ref.json: no $name[]")
    }

    private class View(val id: Int)

    @Suppress("UNCHECKED_CAST")
    private fun runSteps(steps: List<Map<String, Any?>>): List<Any?> {
        val reg = RefRegistry<View>()
        val views = HashMap<Int, View>()
        fun viewFor(n: Int) = views.getOrPut(n) { View(n) }
        return steps.map { s ->
            val v = viewFor((s["view"] as? Number)?.toInt() ?: 1)
            when {
                s["appear"] != null -> { reg.provide(s["appear"] as String, v); null }
                s["disappear"] != null -> { reg.clear(s["disappear"] as String, v); null }
                s["collect"] != null -> { reg.collect(s["collect"] as String); null }
                s["resolve"] != null -> if (reg.resolve(s["resolve"] as String) != null) "live" else StackRef.UNKNOWN
                s["resolveView"] != null -> reg.resolve(s["resolveView"] as String)?.id
                else -> error("step names no operation")
            }
        }
    }

    @Test
    fun keyDerivationAgreesWithCorpus() {
        for (case in section("keys")) {
            val name = case["name"] as? String ?: "<unnamed>"
            assertEquals(case["expect"] as String?, StackRef.key(case["ref"] as? String), name)
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun lifecycleAgreesWithCorpus() {
        for (case in section("lifecycle")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = (case["expect"] as List<Any?>).map { it }
            assertEquals(expect, runSteps(case["steps"] as List<Map<String, Any?>>), name)
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun recyclingAgreesWithCorpus() {
        for (case in section("reuse")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = (case["expect"] as List<Any?>).map { (it as? Number)?.toInt() ?: it }
            assertEquals(expect, runSteps(case["steps"] as List<Map<String, Any?>>), name)
        }
    }
}

File: Web/packages/kernel/test/ref-conformance.test.ts
//
//  ref-conformance.test.ts — the SHARED `ref=` corpus through the TS kernel
//  (OpenSource/Conformance/input/ref.json). The Kotlin twin (:core RefConformanceTest) and the
//  Swift twin (StackRef) run the SAME file.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import { refKey, RefRegistry } from "../src/index.ts";

function corpus(): { [k: string]: unknown } {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const c = join(dir, "OpenSource/Conformance/input/ref.json");
    if (existsSync(c)) {
      const doc = JSON.parse(readFileSync(c, "utf-8")) as { [k: string]: unknown };
      assert.equal(doc["version"], 1, "ref.json: version");
      return doc;
    }
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/input/ref.json not found");
    dir = parent;
  }
}

type Step = { appear?: string; disappear?: string; resolve?: string; resolveView?: string; collect?: string; view?: number };

function runSteps(steps: Step[]): (string | number | null)[] {
  const reg = new RefRegistry<object>();
  const views = new Map<number, object>();
  const viewFor = (n: number): object => {
    if (!views.has(n)) views.set(n, { id: n });
    return views.get(n)!;
  };
  const idOf = (v: object): number => (v as { id: number }).id;
  return steps.map((s) => {
    if (s.appear !== undefined) { reg.provide(s.appear, viewFor(s.view ?? 1)); return null; }
    if (s.disappear !== undefined) { reg.clear(s.disappear, viewFor(s.view ?? 1)); return null; }
    if (s.collect !== undefined) { reg.collect(s.collect); return null; }
    if (s.resolve !== undefined) { const r = reg.resolve(s.resolve); return r.ok ? "live" : r.error; }
    if (s.resolveView !== undefined) { const r = reg.resolve(s.resolveView); return r.ok ? idOf(r.view) : null; }
    throw new Error("step names no operation");
  });
}

test("ref: key derivation agrees with the corpus", () => {
  const cases = corpus()["keys"] as { name: string; ref?: string; expect: string | null }[];
  assert.ok(cases.length > 0);
  for (const c of cases) assert.equal(refKey(c.ref), c.expect, c.name);
});

test("ref: the provider lifecycle agrees with the corpus", () => {
  const cases = corpus()["lifecycle"] as { name: string; steps: Step[]; expect: (string | null)[] }[];
  assert.ok(cases.length > 0);
  for (const c of cases) assert.deepEqual(runSteps(c.steps), c.expect, c.name);
});

test("ref: the recycling rule agrees with the corpus", () => {
  const cases = corpus()["reuse"] as { name: string; steps: Step[]; expect: (string | number | null)[] }[];
  assert.ok(cases.length > 0);
  for (const c of cases) assert.deepEqual(runSteps(c.steps), c.expect, c.name);
});
